hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/TranslateToPublicAddressProvider.java

```diff
@@ -0,0 +1,133 @@
+    private static final EndpointQualifier CLIENT_PUBLIC_ENDPOINT_QUALIFIER =
+            EndpointQualifier.resolve(ProtocolType.CLIENT, "public");
+    private final ClientConfig config;
+    private final HazelcastProperties properties;
+    private final ILogger logger;
+
+    private volatile boolean translateToPublicAddress;
+
+    TranslateToPublicAddressProvider(ClientConfig config, HazelcastProperties properties, ILogger logger) {
+        this.config = config;
+        this.properties = properties;
+        this.logger = logger;
+    }
+
+    void refresh(AddressProvider addressProvider, Collection<MemberInfo> members) {
+        translateToPublicAddress = resolve(addressProvider, members);
+    }
+
+    private boolean resolve(AddressProvider addressProvider, Collection<MemberInfo> members) {
+        if (!(addressProvider instanceof DefaultAddressProvider)) {
+            return false;
+        }
+
+        // Default value of DISCOVERY_SPI_PUBLIC_IP_ENABLED is `null` intentionally.
+        // If DISCOVERY_SPI_PUBLIC_IP_ENABLED is not set to true/false, we don't know the intention of the user,
+        // we will try to decide if we should use private/public address automatically in that case.
+        String publicIpEnabledProperty = properties.getString(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED);
+        if (publicIpEnabledProperty == null) {
+            if (members.isEmpty() || memberInternalAddressAsDefinedInClientConfig(members)) {
+                return false;
+            }
+
+            return membersReachableOnlyViaPublicAddress(members);
+        }
+        return properties.getBoolean(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED);
+    }
+
+    /**
+     * Checks if any member has its internal address as configured in ClientConfig.
+     * <p>
+     * If any member has its internal/private address the same as configured in ClientConfig, then it means that the client is
+     * able to connect to members via configured address. No need to use make any address translation.
+     */
+    boolean memberInternalAddressAsDefinedInClientConfig(Collection<MemberInfo> members) {
+        List<String> addresses = config.getNetworkConfig().getAddresses();
+        return members.stream()
+                .map(MemberInfo::getAddress)
+                .anyMatch(a -> addresses.contains(a.getHost())
+                        || addresses.contains(String.format("%s:%s", a.getHost(), a.getPort())));
+    }
+
+    /**
+     * Checks if members are reachable via public addresses, but not reachable via internal addresses.
+     * <p>
+     * We check only limited number of random members to reduce the slowdown of the startup.
+     */
+    private boolean membersReachableOnlyViaPublicAddress(Collection<MemberInfo> members) {
+        Iterator<MemberInfo> iter = members.iterator();
+        for (int i = 0; i < REACHABLE_CHECK_NUMBER; i++) {
+            if (!iter.hasNext()) {
+                iter = members.iterator();
```

Comment:
@leszko It looks like we are retrying some members twice if the number of members is less than `REACHABLE_CHECK_NUMBER`.  If so, I think we should not. Am I missing something?

Reply:
My reasoning was that, for example, if we have just 1 member, we should anyway retry a few times, because otherwise, we risk that there was some random network congestion for just one moment and we take a decision based on that. So, in my opinion, I would retry a few times even if we only have 1 member.